sources/core/Stride.Core.CompilerServices/Analyzers/STRDIAG005ReadonlyMemberIsReferenceType.cs

```diff
@@ -0,0 +1,89 @@
+    private const string Title = "ReadOnlyMemberIsReferenceType";
+    private const string MessageFormat = "The [DataMember] Attribute is applied to a read-only member '{0}' with a non allowed Type. Use a reference type instead, string isn't allowed either.";
+    private const string Category = "CompilerServices";
+
+    private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+        DiagnosticId,
+        Title,
+        MessageFormat,
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.EnableConcurrentExecution();
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+        context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+    }
+    private static void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+    {
+        var dataMemberAttribute = WellKnownReferences.DataMemberAttribute(context.Compilation);
+        if (dataMemberAttribute is null)
+        {
+            return;
+        }
+
+        context.RegisterSymbolAction(symbolContext => AnalyzeProperty(symbolContext, dataMemberAttribute), SymbolKind.Property);
+        context.RegisterSymbolAction(symbolContext => AnalyzeField(symbolContext, dataMemberAttribute), SymbolKind.Field);
+    }
+    private static void AnalyzeField(SymbolAnalysisContext context,INamedTypeSymbol dataMemberAttribute)
+    {
+        var symbol = context.Symbol;
+
+        if (symbol is IFieldSymbol fieldSymbol)
+        {
+            if (fieldSymbol.GetAttributes().Any(attr => attr.AttributeClass?.Equals(dataMemberAttribute, SymbolEqualityComparer.Default) ?? false) &&
+                fieldSymbol.IsReadOnly)
+            {
+                var fieldType = fieldSymbol.Type;
+                if (fieldType is null)
+                    return;
+                if (fieldType.SpecialType == SpecialType.System_String || !fieldType.IsReferenceType)
+                {
+                    var diagnostic = Diagnostic.Create(Rule, fieldSymbol.Locations.First(), fieldSymbol.Name, fieldType);
+                    context.ReportDiagnostic(diagnostic);
+                }
+            }
+        }
+    }
+    private static void AnalyzeProperty(SymbolAnalysisContext context, INamedTypeSymbol dataMemberAttribute)
+    {
+        var propertySymbol = (IPropertySymbol)context.Symbol;
+        if (!WellKnownReferences.HasAttribute(propertySymbol, dataMemberAttribute))
+            return;
+
+        var propertyType = propertySymbol.Type;
+        if (propertyType is null)
+            return;
+
+        if (propertySymbol.GetMethod != null && (propertyType.SpecialType == SpecialType.System_String || !propertyType.IsReferenceType))
```

Comment:
Can you reorganize those if statements a little?
First check that method is read-only and only then check the type.
Also, maybe extract the type check to another method shared by both property and field analyzers?

Reply:
is rewritten to an easier way to understand it